### PR TITLE
[RouteFactory] Passing generic type to `RouteFactory`

### PR DESCRIPTION
**Pull request by AlexV525**
## Description

This will solve the problem when passing type argument through `onGenerateRoute` but only `dynamic` can be returned.
Tests that affected by this changes have been updated. But some cases such as `removeRoute` test in `navigator_test` definitely needs to be change.

### Example

Please try this with https://github.com/AlexVincent525/flutter_test_app , in `Test generic type route page` I've provided a method with the `push` button that wants `pushNamed` return a boolean type of result. It failed on master because `onGenerateRoute` doesn't carried `T` with it.

## Related Issues
 
Might be related to #50629 .

## Tests

Add `Passing generic type to onGenerateRoute` test case in `navigator_test`.

## Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] All existing and new tests are passing.
- [ ] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

I'm not sure whether it's a breaking change, but after this, various of test case that implemented `onGenerateRoute` needs to be update, so I think it might be.

- [ ] No, no existing tests failed, so this is *not* a breaking change.
- [x] Yes, this is a breaking change.
   - [ ] I wrote a design doc: https://flutter.dev/go/template 
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [x] I wrote a migration guide.

### CODE REVIEW

**Comment by AlexVincent525** [comment]
cc @HansMuller

**Comment by AlexVincent525** [comment]
Seems some code were changed by `dartfmt`, plz ignore them if they're not necessary to be fixed.

**Comment by AlexVincent525** [comment]
Please help me review this PR when you guys have time. ❤️ @HansMuller @dnfield @xster

File: packages/flutter/test/cupertino/nav_bar_test.dart [hunk 1]
```diff
@@ -928,8 +928,8 @@ void main() {
     await tester.pumpWidget(
       WidgetsApp(
         color: const Color(0xFFFFFFFF),
-        onGenerateRoute: (RouteSettings settings) {
-          return CupertinoPageRoute<void>(
+        onGenerateRoute: <T>(RouteSettings settings) {
```
Comment (dnfield) [inline packages/flutter/test/cupertino/nav_bar_test.dart]: This is a breaking change and will need to follow our policy on making breaking changes.

**Comment by dnfield** [comment]
Please revert any dartfmt related changes - we don't use dartfmt in this repository.

You also have some merge conflicts to resolve.  

This will require at _least_ a migration guide.  I'm fairly certain this will also require internal changes in Google's repo.

@chunhtai - how does this fit in with other planned changes for navigator?

/cc @goderbauer

**Comment by AlexVincent525** [comment]
> Please revert any dartfmt related changes - we don't use dartfmt in this repository.
> 
> You also have some merge conflicts to resolve.

Will done these ASAP.

**Comment by AlexVincent525** [comment]
> This will require at _least_ a migration guide. I'm fairly certain this will also require internal changes in Google's repo.

Is there a migration guide I can follow? Or I need to wait for your guys feel good to process with this PR?

**Comment by dnfield** [comment]
There are a number of examples at https://flutter.dev/docs/release/breaking-changes - the git repo for that is https://github.com/flutter/website.git

**Comment by AlexVincent525** [comment]
Too many 'revert' commit for now, I'll open up a new PR for this.

**Comment by AlexVincent525** [comment]
> There are a number of examples at https://flutter.dev/docs/release/breaking-changes - the git repo for that is https://github.com/flutter/website.git

Done with migration guide.

**Comment by goderbauer** [comment]
> This will solve to problem when passing type argument through onGenerateRoute but only dynamic can be returned.

dynamic allows you to return whatever you want. You can cast it to whatever type you like. While some stronger static type safety may be desirable here, I don't understand what problem this is solving. Can you elaborate what this enables you to do that you couldn't do before to desired whether this breaking change is worth it?

**Comment by goderbauer** [comment]
Also: There are many failing tests on cirrus.

**Comment by chunhtai** [comment]
@AlexVincent525 @goderbauer 
see my workaround and repro in https://github.com/flutter/flutter/issues/50629#issuecomment-606130758

**Comment by AlexVincent525** [comment]
> While some stronger static type safety may be desirable here, I don't understand what problem this is solving. Can you elaborate what this enables you to do that you couldn't do before to desired whether this breaking change is worth it?

Let's have a quick look about what methods does `Navigator` have currently, using named route:

`Future<T> pushNamed<T extends Object>`

`Future<T> pushReplacementNamed<T extends Object, TO extends Object>`

`Future<T> popAndPushNamed<T extends Object, TO extends Object>`

`Future<T> pushNamedAndRemoveUntil<T extends Object>`

They all goes to `Route<T> _routeNamed<T>` eventually.

So from my point of view we need to unify the use of these methods with generic type provide cause the base method has generic type carried with.

Imaging that a user wants a fixed type returned when pushing, why he/she needs `as someType` rather than `pushNamed<someType>`? The generic type is design for this scenario.

Another scenario is add-to-app. Using `as someType` will increase the size of apps according to [avoid_as](https://dart-lang.github.io/linter/lints/avoid_as.html) from linter, which they do care about.

While you can only use `as someType` for now, it's really an ugly workaround for a type cast. There's nothing that cannot be done now, but this breaking change means better, though Flutter (I mean official side) should provide a good example about method usage.

> see my workaround and repro in [#50629 (comment)](https://github.com/flutter/flutter/issues/50629#issuecomment-606130758)

Yes it did work, but it's still a **workaround**.

> Also: There are many failing tests on cirrus.

These test cases are updated too fast, I will solve them after the proposal of this PR is approved or it's stable.

**Comment by chunhtai** [comment]
I don't think it make sense to define type at app level. For example, [showdialog](https://api.flutter.dev/flutter/material/showDialog.html) can push page with arbitrary results depends on the use case. This approach does not solve that use case.

**Comment by AlexVincent525** [comment]
> For example, user might have different name route with different return type for custom form response.

This change requires the generic type parameter is passed when define `onGenerateRoute`, please see the migration guide.

**Comment by chunhtai** [comment]
> > For example, user might have different name route with different return type for custom form response.
> 
> This change requires the generic type parameter is passed when define `onGenerateRoute`, please see the migration guide.

No, you are right. I overlooked part of the design

**Comment by chunhtai** [comment]
I have been thinking about what make the most sense. We are actually in the middle of refactoring entire routing system for name route. See [Navigator 2.0](https://docs.google.com/document/d/1Q0jx0l4-xymph9O6zLaOY4d_f7YFpNWX_eGbzYxr9wY/edit#)
and [Router](https://docs.google.com/document/d/1QodTFmwVr1yk8dyQvoqkeILxvm92oikB1bVhHTtp6R4/edit#heading=h.dcnl8p9y9hme)
Once router widget is landed, we will encourage people to migrate to the new system.

I can see this pr fixes some obvious errors. However, a lot of apps will have to migrate after this pr lands. I find it a bit uncertain to make people migrate their apps to an API that will become legacy in the future and they have to migrate again.

Since there is workaround. I suggest we hold off this fix. @AlexVincent525 How do you think?

**Comment by AlexVincent525** [comment]
> Once router widget is landed, we will encourage people to migrate to the new system.

Will this widget land in next stable release or next quarter? If not I think we can continue to push the implementation, since it has been four months when last stable (major version) release. According the release recently with hotfix, I suspect that this won't land until next year :P

> I find it a bit uncertain to make people migrate their apps to an API that will become legacy in the future and they have to migrate again.

Althought it breaks most cases related to navigator currently, the migration can be done easily, which is significant helpful for apps and packages that rely on named route.

**Comment by dnfield** [comment]
We would have to come up with a way to make this a soft break.  There are a large number of users who will experience this as a difficult breakage because of packages they don't control getting broken.  

The next stable is coming soonish.  But either way we should treat everything we commit to master as if it is ready for stable tomorrow.

**Comment by AlexVincent525** [comment]
@chunhtai About your workaround, is it possible to using push<T> rather than `as T`?

**Comment by AlexVincent525** [comment]
During lunch break I quickly browsed the API documentation based on the master branch, and found nothing related to the new navigator or the new router, when will this plan start to deploy? Again if it will take a long term I suggest that we can push this implementation land.
image

**Comment by goderbauer** [comment]
I left a comment on https://github.com/flutter/flutter/issues/50629#issuecomment-608571875 to show another way that the issue can be fixed in user code.

Thinking about this more, I don't think it is appropriate to give onGenerateRoute a generic because by design the method needs to return Routes with different generic types. Let's say in my app I have two routes, one returns the date that a user has selected in a datepicker and another one returns true/false to indicate whether the user wants to perform a certain action. I would defines these routes as follows:

```dart
class DatePickerRoute extends Route<DateTime> {
  // implementation of Route that returns a selected DateTime.
}

class ConfirmDialogRoute extends Route<bool> {
  // implementation of Route that returns whether the user wants to perform an action.
}
```

If onGenerateRoute now takes a generic T, I wouldn't be able to return any of those Routes from my onGenerateRoute because both of them are not of type `Route<T>`. The analyzer would complain:

```
error: A value of type 'DatePickerRoute' can't be returned from function 'onGenerateRoute' because it has a return type of 'Route<T>'. (return_of_invalid_type at <>)
error: A value of type 'ConfirmDialogRoute' can't be returned from function 'onGenerateRoute' because it has a return type of 'Route<T>'. (return_of_invalid_type at <>)
```

**Comment by AlexVincent525** [comment]
(Currently not working, will be AFK for 1 or 2 days.)

**Comment by AlexVincent525** [comment]
Sorry for the delay. Just back to work today.

> If onGenerateRoute now takes a generic T, I wouldn't be able to return any of those Routes from my onGenerateRoute because both of them are not of type `Route<T>`.

I try this with the same error occurred, and by using `as Route<T>` this error solved. But that comes to the origin problem, whether we use `as` or generic type there must be somewhere that should force casting a route to a specific type. That's kinda chaos.

```dart
class DatePickerRoute extends Route<DateTime> { }

onGenerateRoute: <T>(RouteSettings settings) {
  return DatePickerRoute() as Route<T>; // It only worked with force casting.
},
```

**Comment by AlexVincent525** [comment]
Thought about the type force convert again...users **MUST known the exactly type** when they're trying to generate (or push) a new route, that means that no matter whatever the type they specified as defined, it should be as the same type as the generic type eventually, otherwise it'll throw an error. I'm not sure what kind of design proposal is in here by official team.

**Comment by Peng-Qian** [comment]
why not just change

```dart
typedef RouteFactory = Route<dynamic>? Function(RouteSettings settings);
```
to
```dart
typedef RouteFactory = Route<T>? Function(RouteSettings settings);
```

and then, we can pass the generic type to screen and enforce the return type from that screen

Some screens can have a generic type to reduce duplicate code, like SearchScreen<Address> and SearchScreen<PhoneNumber>, but now we can't even pass the generic type via `Navigator.of(context).pushNamed`

**Pull request closed without merging**